apps-rendering/src/server/insertNewsletterIntoItem.ts

```diff
@@ -0,0 +1,306 @@
+import { ElementKind } from 'bodyElementKind';
+import type { Item } from 'item';
+import { logger } from 'logger';
+import { Result } from 'result';
+import { stringToPillar } from 'themeStyles';
+
+// ----- Constants ----- //
+
+const enum ElementCategory {
+	'ParagraphText' = 'PARAGRAPH',
+	'BoldParagraphText' = 'BOLD_PARAGRAPH',
+	'NonParagraphText' = 'OTHER TEXT ELEMENT',
+	'Heading' = 'HEADING',
+	'NonText' = 'NON TEXT',
+	'WhiteSpace' = 'WHITE SPACE',
+	'Error' = 'ERROR',
+}
+
+const DEBUG = {
+	log: false,
+	logContent: false,
+};
+
+// ----- pure functions ---//
+
+function categoriseElement(
+	result: Result<string, BodyElement>,
+): ElementCategory {
+	if (!result.isOk()) {
+		// using result.isErr does not convince the compiler that result.value exists
+		return ElementCategory.Error;
+	}
+
+	switch (result.value.kind) {
+		case ElementKind.HeadingThree:
+		case ElementKind.HeadingTwo:
+			return ElementCategory.Heading;
+		case ElementKind.Text: {
+			const { doc } = result.value;
+			if (doc.textContent?.trim().length === 0) {
+				return ElementCategory.WhiteSpace;
+			}
+
+			if (doc.nodeName === 'P') {
+				if (doc.nodeType === 1) {
+					const element = doc as Element;
+					// for paragraphs containing only single <strong> child ("bold text"),
+					// the sign up block can be put before them, but not after them.
+					if (
+						element.children.length === 1 &&
+						element.firstElementChild?.tagName === 'STRONG'
+					) {
+						return ElementCategory.BoldParagraphText;
+					}
+				}
+				return ElementCategory.ParagraphText;
+			} else {
+				return ElementCategory.NonParagraphText;
+			}
+		}
+		default:
+			return ElementCategory.NonText;
+	}
+}
+
+/**
+ * Get the target range of places within a list of visible elements
+ * that the sign-up block can be placed in, being the places within
+ * 3 elements from the middle of the article (by element number only;
+ * not taking account of the length/height of each element)
+ *
+ * @param numberOfElements
+ * @returns The minimum and maximum index for the target range
+ */
+function getRange(numberOfElements: number): [number, number] {
+	const middle = Math.floor(numberOfElements / 2);
+	return [Math.max(0, middle - 3), Math.min(numberOfElements, middle + 3)];
+}
+
+/**
+ * Predict the indices of bodyElement that would
+ * have an ad slot inserted before them on the current
+ * ad placement logic.
+ *
+ * This logic is not perfect as the ad placement logic
+ * insert after the bodyElement have been renderer by
+ * counting paragraph elements in ReactNodes, whereas here
+ * we are looking at bodyElement before they are have been
+ * rendered.
+ *
+ * @param contentOnlyBody an subset of an article Body, with the
+ * non-rendered elements (errors and whitespace) removed.
+ * @returns the indices of the elements that are expected to
+ * have an adslot placed before them.
+ */
+function predictIndicesOfParagraphsAfterAdslots(
```

Comment:
It is a hard requirement, but not once that necessarily needs to be implemented in this PR - the components aren't going live yet, so we could remove the logic and look separately at how to  address ads.

Currently, the ads are placed after the `BodyElements` have been rendered to `ReactNodes`. I don't think this kind of logic would work well at stage because we've then lost the structure from the data model.  Since [#5767](https://github.com/guardian/dotcom-rendering/pull/5767), there are `BodyElements` for individual paragraphs, so it might be moving the ad placement to be before the render to `ReactNodes` (and then have placing Sign Ups after placing ads, but both before the React stage).

Reply:
yes, that makes sense. We had a discussion this morning about how to take this forward and think it's a good idea to:

- keep the newsletter signup placement logic here
- if necessary, teach ads how to avoid newsletter signups for now
- later, look into refactoring ad placement to also happen pre-render

Reply:
Great - I've removed the 'predict ads' part. It would be good to pair on changes to the ads module.